docs/features/incremental-generators.md

```diff
@@ -0,0 +1,478 @@
+IncrementalValueSource<(AdditionalText source1Item, IEnumerable<Compilation> source2Batch)> combined = context.Sources.AdditionalTexts.Combine(context.Sources.Compilation);
+```
+
+The type of combined is an
+`IncrementalValueSource<(AdditionalText, IEnumerable<Compilation>)>`. For each
+additional text, there is a tuple with the text, and the batched data of the
+compilation source. As the author knows that there is only ever a single
+compilation, they are free to transform the data to select the single
+compilation object:
+
+```csharp
+IncrementalValueSource<(AdditionalText, Compilation)> transformed = combined.Transform(static pair => (pair.source1Item, pair.source2Batch.Single()));
+```
+
+Similarly a cross join can be achieved by first combining two value sources,
+then batch transforming the resulting value source.
+
+```csharp
+IncrementalValueSource<(AdditionalText, MetadataReference)> combined = context.Sources.AdditionalTexts
+                                                                                      .Combine(context.Sources.MetadataReference)
+                                                                                      .TransformMany(static pair => pair.source2Batch.Select(static metadataRef => (pair.source1Item, metadataRef));
+```
+
+**OPEN QUESTION**: Combine is pretty low level, but does allow you to do
+everything needed when used in conjunction with transform. Should we provide
+some higher level extension methods that chain together combine and transform
+out of the box for common operations?
+
+While filtering can be easily enough implement by the user as a transform step,
+it seems common and useful enough that we provide an implementation directly for
+the user to consume
+
+```csharp
+static partial class IncrementalValueSourceExtensions
+{
+    // helper for filtering values
+    public static IncrementalValueSource<T> Filter<T>(this IncrementalValueSource<T> source, Func<T, bool> filter) => ...
+}
+```
+
+### Caching
+
+While the finer grained steps allow for some coarse control of output types via
+the generator host, the performance benefits are only really seen when the
+driver can cache the outputs from one pipeline step to the next. While we have
+generally said that the execute method in an `ISourceGenerator` should be
+deterministic, incremental generators actively _require_ this property to be
+true.
+
+When calculating the required transformations to be applied as part of a step,
+the generator driver is free to look at inputs it has seen before and used
+previous computed and cached values of the transformation for these inputs. When
+using non batch transforms it can do this on an element by element basis.
+
+Consider the following transform:
+
+```csharp
+IValueSource<string> transform = context.Sources.AdditionalTexts
+                                                .Transform(static t => t.Path)
+                                                .Transform(static p => "prefix_" + p);
```

Comment:
nit: although I understand this is for illustration purpose, I'm a bit nervous about using such trivial steps and encouraging caching trivial computations. 
For follow-up: we'll want to provide some guidance on granularity of steps.